### PR TITLE
add funcvar

### PR DESCRIPTION
**Pull request by jumormt**

### CODE REVIEW

**Comment by jumormt** [comment]
--- log/redis-server.log	2024-11-23 22:24:42.786280885 +1100
+++ log/redis-server-xiao.log	2024-11-23 22:26:20.009204378 +1100
@@ -31,9 +31,9 @@
 VarArrayObj         831
 VarStructObj        1664
 ----------------Time and memory stats--------------------
-LLVMIRTime          4.304
-SVFIRTime           2.52
-SymbolTableTime     0.572
+LLVMIRTime          4.761
+SVFIRTime           2.712
+SymbolTableTime     0.642
 #######################################################
 
 *********PTACallGraph Stats (Andersen analysis)***************
@@ -60,11 +60,11 @@
 CollapseTime        0
 CopyGepTime         0
 LoadStoreTime       0
-MemoryUsageVmrss    4.16476e+06
+MemoryUsageVmrss    4.16454e+06
 MemoryUsageVmsize   4.16453e+06
 SCCDetectTime       0
 SCCMergeTime        0
-TotalTime           129.268
+TotalTime           129.251
 UpdateCGTime        0
 ----------------Numbers stats----------------------------
 AddrProcessed       17887
@@ -115,32 +115,32 @@
 ****Persistent Points-To Cache Statistics: Andersen's analysis bitvector****
 ################ (program : redis-server.bc)###############
 UniquePointsToSets       44837
-TotalUnions              673050
+TotalUnions              673062
 PropertyUnions           222658
-UniqueUnions             19260
-LookupUnions             395148
-PreemptiveUnions         35984
+UniqueUnions             19262
+LookupUnions             395146
+PreemptiveUnions         35996
 TotalComplements         3364152
 PropertyComplements      2998574
 UniqueComplements        16887
 LookupComplements        331806
 PreemptiveComplements    16885
-TotalIntersections       4182801
-PropertyIntersections    4145520
-UniqueIntersections      840
-LookupIntersections      2154
-PreemptiveIntersections  34287
+TotalIntersections       4182811
+PropertyIntersections    4145494
+UniqueIntersections      848
+LookupIntersections      2172
+PreemptiveIntersections  34297
 #######################################################
 
 *********Memory SSA Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
 AverageRegSize      85.727
-GenMUCHITime        1.853
-GenRegionTime       462.748
-InsertPHITime       0.547
-SSARenameTime       0.09
-TotalMSSATime       465.248
+GenMUCHITime        1.966
+GenRegionTime       513.809
+InsertPHITime       0.585
+SSARenameTime       0.083
+TotalMSSATime       516.448
 ----------------Numbers stats----------------------------
 BBHasMSSAPhi        10755
 CSChiNode           111246
@@ -163,13 +163,13 @@
 *********SVFG Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-ATNodeTime          0.846
+ATNodeTime          0.928
 AvgWeight           167.789
 ConnDirEdgeTime     0
-ConnIndEdgeTime     4.43
+ConnIndEdgeTime     4.47
 OptTime             0
 TLNodeTime          0
-TotalTime           5.276
+TotalTime           5.398
 ----------------Numbers stats----------------------------
 ActualIn            163193
 ActualOut           111246
@@ -231,20 +231,20 @@
 GepTime             0
 IndirectPropaTime   0
 LoadTime            0
-MemoryUsageVmrss    3.57894e+06
-MemoryUsageVmsize   3.61746e+06
+MemoryUsageVmrss    3.58674e+06
+MemoryUsageVmsize   3.6212e+06
 PhiTime             0
-PrelabelingTime     0.148
+PrelabelingTime     0.139
 ProcessTime         0
 PropagationTime     0
 SCCTime             0
-SolveTime           441.365
+SolveTime           456.631
 StoreTime           0
 Strong/WeakUpdTime  0
-TotalTime           654.298
+TotalTime           696.712
 UpdateCGTime        0
 VersionPropTime     0
-meldLabelingTime    209.556
+meldLabelingTime    236.534
 ----------------Numbers stats----------------------------
 CopysNum            14138
 DummyFieldPtrs      2305
@@ -262,13 +262,13 @@
 Pointers            494739
 ProcessedAParam     0
 ProcessedAddr       107322
-ProcessedCopy       162850
+ProcessedCopy       164271
 ProcessedFRet       0
-ProcessedGep        868341
-ProcessedLoad       874876
+ProcessedGep        879876
+ProcessedLoad       869135
 ProcessedMSSANode   2293638
-ProcessedPhi        104659
-ProcessedStore      321638
+ProcessedPhi        106726
+ProcessedStore      322168
 SolveIterations     6
 StoresNum           24164
 StrongUpdates       3378
@@ -283,12 +283,12 @@
 
 ****Persistent Points-To Cache Statistics: flow-sensitive analysis bitvector****
 ################ (program : redis-server.bc)###############
-UniquePointsToSets       29728
-TotalUnions              1144666386
-PropertyUnions           989064136
-UniqueUnions             38169
-LookupUnions             155512743
-PreemptiveUnions         51338
+UniquePointsToSets       29868
+TotalUnions              1140463970
+PropertyUnions           977053095
+UniqueUnions             38732
+LookupUnions             163320068
+PreemptiveUnions         52075
 TotalComplements         0
 PropertyComplements      0
 UniqueComplements        0

**Comment by yuleisui** [comment]
Funcval=>FunVal

Funcvar=>FunVar

**Comment by yuleisui** [comment]
Time stats have also increased significantly

**Comment by jumormt** [comment]
> Time stats have also increased significantly

Ok, I'll check.

**Comment by jumormt** [comment]
I haven't merged the latest SVF master. I just reran both versions and the time stat is normal now:
--- log/redis-server.log	2024-11-23 01:47:11.652411669 +1100
+++ log/redis-server-xiao.log	2024-11-23 01:47:52.907989433 +1100
@@ -31,9 +31,9 @@
 VarArrayObj         831
 VarStructObj        1664
 ----------------Time and memory stats--------------------
-LLVMIRTime          4.257
-SVFIRTime           2.476
-SymbolTableTime     0.554
+LLVMIRTime          4.316
+SVFIRTime           2.517
+SymbolTableTime     0.55
 #######################################################
 
 *********PTACallGraph Stats (Andersen analysis)***************
@@ -60,11 +60,11 @@
 CollapseTime        0
 CopyGepTime         0
 LoadStoreTime       0
-MemoryUsageVmrss    4.16477e+06
+MemoryUsageVmrss    4.16465e+06
 MemoryUsageVmsize   4.16454e+06
 SCCDetectTime       0
 SCCMergeTime        0
-TotalTime           125.897
+TotalTime           128.428
 UpdateCGTime        0
 ----------------Numbers stats----------------------------
 AddrProcessed       17887
@@ -115,32 +115,32 @@
 ****Persistent Points-To Cache Statistics: Andersen's analysis bitvector****
 ################ (program : redis-server.bc)###############
 UniquePointsToSets       44837
-TotalUnions              673050
+TotalUnions              673064
 PropertyUnions           222658
-UniqueUnions             19263
-LookupUnions             395145
-PreemptiveUnions         35984
+UniqueUnions             19259
+LookupUnions             395149
+PreemptiveUnions         35998
 TotalComplements         3364152
 PropertyComplements      2998574
 UniqueComplements        16887
 LookupComplements        331806
 PreemptiveComplements    16885
-TotalIntersections       4182798
-PropertyIntersections    4145543
-UniqueIntersections      834
-LookupIntersections      2137
-PreemptiveIntersections  34284
+TotalIntersections       4182816
+PropertyIntersections    4145478
+UniqueIntersections      852
+LookupIntersections      2184
+PreemptiveIntersections  34302
 #######################################################
 
 *********Memory SSA Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
 AverageRegSize      85.727
-GenMUCHITime        1.877
-GenRegionTime       454.979
-InsertPHITime       0.555
+GenMUCHITime        1.932
+GenRegionTime       490.242
+InsertPHITime       0.582
 SSARenameTime       0.086
-TotalMSSATime       457.501
+TotalMSSATime       492.852
 ----------------Numbers stats----------------------------
 BBHasMSSAPhi        10755
 CSChiNode           111246
@@ -163,13 +163,13 @@
 *********SVFG Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-ATNodeTime          0.856
+ATNodeTime          0.87
 AvgWeight           167.789
 ConnDirEdgeTime     0
-ConnIndEdgeTime     4.468
+ConnIndEdgeTime     4.475
 OptTime             0
 TLNodeTime          0
-TotalTime           5.324
+TotalTime           5.345
 ----------------Numbers stats----------------------------
 ActualIn            163193
 ActualOut           111246
@@ -231,20 +231,20 @@
 GepTime             0
 IndirectPropaTime   0
 LoadTime            0
-MemoryUsageVmrss    3.58736e+06
-MemoryUsageVmsize   3.61802e+06
+MemoryUsageVmrss    3.57212e+06
+MemoryUsageVmsize   3.61089e+06
 PhiTime             0
-PrelabelingTime     0.147
+PrelabelingTime     0.138
 ProcessTime         0
 PropagationTime     0
 SCCTime             0
-SolveTime           445.018
+SolveTime           448.373
 StoreTime           0
 Strong/WeakUpdTime  0
-TotalTime           659.41
+TotalTime           660.91
 UpdateCGTime        0
 VersionPropTime     0
-meldLabelingTime    211.043
+meldLabelingTime    209.111
 ----------------Numbers stats----------------------------
 CopysNum            14138
 DummyFieldPtrs      2305
@@ -262,13 +262,13 @@
 Pointers            494739
 ProcessedAParam     0
 ProcessedAddr       107322
-ProcessedCopy       165092
+ProcessedCopy       163188
 ProcessedFRet       0
-ProcessedGep        874801
-ProcessedLoad       873234
+ProcessedGep        867312
+ProcessedLoad       855572
 ProcessedMSSANode   2293638
-ProcessedPhi        105845
-ProcessedStore      321643
+ProcessedPhi        104237
+ProcessedStore      317338
 SolveIterations     6
 StoresNum           24164
 StrongUpdates       3378
@@ -283,12 +283,12 @@
 
 ****Persistent Points-To Cache Statistics: flow-sensitive analysis bitvector****
 ################ (program : redis-server.bc)###############
-UniquePointsToSets       29652
-TotalUnions              1169393112
-PropertyUnions           1005513092
-UniqueUnions             37998
-LookupUnions             163790900
-PreemptiveUnions         51122
+UniquePointsToSets       29453
+TotalUnions              1122736818
+PropertyUnions           971251998
+UniqueUnions             36970
+LookupUnions             151397999
+PreemptiveUnions         49851
 TotalComplements         0
 PropertyComplements      0
 UniqueComplements        0

File: svf/include/SVFIR/SVFIR.h [hunk 1]
```diff
@@ -545,26 +545,36 @@ class SVFIR : public IRGraph
...
     }
+
+    NodeID addFuncValNode(const CallGraphNode* callGraphNode, NodeID i, const ICFGNode* icfgNode) {
+        FuncValVar* node = new FuncValVar(callGraphNode, i, icfgNode);
+        return addValNode(nullptr, node, i);
+    }
+
     /// Add a memory obj node
-    inline NodeID addObjNode(const SVFValue* val, NodeID i)
+    inline NodeID addObjNode(const SVFValue* val, const CallGraphNode* callGraphNode, NodeID i)
```
Comment (jumormt) [inline svf/include/SVFIR/SVFIR.h:560]: split into multiple functions

**Pull request merged**